src/journal-remote/journal-gatewayd.c

```diff
@@ -283,70 +296,149 @@ static int request_parse_accept(
         return 0;
 }
 
-static int request_parse_range(
+static int request_parse_range_skip_and_n_entries(
                 RequestMeta *m,
-                struct MHD_Connection *connection) {
+                const char *colon) {
 
-        const char *range, *colon, *colon2;
+        const char *p, *colon2;
         int r;
 
-        assert(m);
-        assert(connection);
+        colon2 = strchr(colon + 1, ':');
+        if (colon2) {
+                _cleanup_free_ char *t;
 
-        range = MHD_lookup_connection_value(connection, MHD_HEADER_KIND, "Range");
-        if (!range)
-                return 0;
+                t = strndup(colon + 1, colon2 - colon - 1);
+                if (!t)
+                        return -ENOMEM;
 
-        if (!startswith(range, "entries="))
-                return 0;
+                r = safe_atoi64(t, &m->n_skip);
+                if (r < 0)
+                        return r;
+        }
 
-        range += 8;
-        range += strspn(range, WHITESPACE);
+        p = (colon2 ? colon2 : colon) + 1;
+        if (*p) {
+                r = safe_atou64(p, &m->n_entries);
+                if (r < 0)
+                        return r;
 
-        colon = strchr(range, ':');
-        if (!colon)
-                m->cursor = strdup(range);
-        else {
-                const char *p;
+                if (m->n_entries <= 0)
+                        return -EINVAL;
 
-                colon2 = strchr(colon + 1, ':');
-                if (colon2) {
-                        _cleanup_free_ char *t;
+                m->n_entries_set = true;
+        }
 
-                        t = strndup(colon + 1, colon2 - colon - 1);
-                        if (!t)
-                                return -ENOMEM;
+        return 0;
+}
 
-                        r = safe_atoi64(t, &m->n_skip);
-                        if (r < 0)
-                                return r;
-                }
+static int request_parse_range_entries(
+                RequestMeta *m,
+                const char *range) {
 
-                p = (colon2 ? colon2 : colon) + 1;
-                if (*p) {
-                        r = safe_atou64(p, &m->n_entries);
-                        if (r < 0)
-                                return r;
+        const char *colon;
+        int r;
 
-                        if (m->n_entries <= 0)
-                                return -EINVAL;
+        range += 8;
+        range += strspn(range, WHITESPACE);
 
-                        m->n_entries_set = true;
-                }
+        colon = strchr(range, ':');
+        if (!colon) {
+                m->cursor = mfree(m->cursor);
```

Comment:
Please move this `mfree()` before `colon = strchr(range, ':');` Otherwise, `m->cursor` may leak in the `else` branch.

Reply:
Done